AutoGraph/WebSocketClient.swift

```diff
@@ -0,0 +1,119 @@
+import Foundation
+import Starscream
+
+open class WebSocketClient {
+    public let baseUrl: String
+    public var httpHeaders: [String : String]
+    public var webSocket: WebSocket?
+    public var lifeCycle: GlobalLifeCycle?
+    private var webSocketListener: WebSocketListener?
+    
+    public init(baseUrl: String,
+                httpHeaders: [String : String] = [:]) {
+        self.baseUrl = baseUrl
+        self.httpHeaders = httpHeaders
+    }
+    
+    open func send<R: Request>(_ request: R, completion: @escaping RequestCompletion<R.SerializedObject>) {
+        do {
+            let query = try request.queryDocument.graphQLString()
+            var parameters: [String : Any] = ["query" : query]
+            if let variables = try request.variables?.graphQLVariablesDictionary() {
+                parameters["variables"] = variables
+            }
+            
+            guard let url = URL(string: self.baseUrl) else {
+                return
+            }
+            
+            let request = try URLRequest(url: url, method: .post)
+            
+            self.webSocket = WebSocket(request: request)
+            self.handleWebSocketListener(with: completion)
+            self.webSocket?.delegate = self.webSocketListener
+        }
+        catch let error {
+            completion(.failure(error))
+        }
+    }
+    
+    public func setAuthToken(_ token: String) {
+        self.httpHeaders["Authorization"] = "Bearer \(token)"
+    }
+    
+    public func disconnect() {
+        self.webSocket?.disconnect()
+        self.webSocket = nil
+        self.webSocketListener = nil
+    }
+    
+    private func handleWebSocketListener<SerializedObject: Decodable>(with completion: @escaping RequestCompletion<SerializedObject>) {
+        self.webSocketListener = WebSocketListener { (action) in
+            switch action {
+            case .connected:
+                break
+            case let .data(data):
+                do {
+                    let decoder = JSONDecoder()
+                    let object = try decoder.decode(SerializedObject.self, from: data)
+                    completion(.success(object))
+                }
+                catch let error {
+                    completion(.failure(error))
+                }
+            case .cancelled:
+                self.disconnect()
+            case let .error(error):
+                if let error = error {
+                    completion(.failure(error))
+                }
+            case .disconnected:
+                self.disconnect()
+            }
+        }
+    }
+}
+
+class WebSocketListener {
+    enum Action {
+        case connected([String: String])
+        case disconnected(String, UInt16)
+        case data(Data)
+        case error(Error?)
+        case cancelled
```

Comment:
so far I'm not seeing a use case for our own set of events here

Reply:
Okay will see how I can integrate that